src/hooks/http.ts

```diff
@@ -155,6 +156,69 @@ function shouldIgnoreRequest(request: http.IncomingMessage) {
   return false;
 }
 
+type Chunk = string | Buffer;
+
+function captureResponseBody(response: http.ServerResponse, chunks: Chunk[]) {
+  let currentLength = 0;
+  function captureChunk(chunk: Chunk) {
+    if (!chunk) return;
+    if (config.responseBodyMaxLength <= currentLength) return;
+
+    chunks.push(chunk);
+    currentLength += chunk.length;
+  }
+
+  const originalWrite = response.write.bind(response);
+  const originalEnd = response.end.bind(response);
+
+  response.write = function (...args: unknown[]) {
+    let callback: unknown;
+    let encoding: unknown;
+    const chunk = args[0] as Buffer;
+    // write has these call signatures
+    // (chunk, callback?)
+    // (chunk, encoding, callback?)
+    let argsToConsider = args.length;
+    // If the last parameter is a function than it's the callback
+    if (args.length > 0 && typeof args[args.length - 1] === "function") {
+      callback = args[args.length - 1];
+      argsToConsider--;
+    }
+    if (argsToConsider > 1) encoding = args[1];
+
+    captureChunk(chunk);
+
+    if (encoding != null)
+      return originalWrite(
+        chunk,
+        encoding as BufferEncoding,
+        callback as ((error: Error | null | undefined) => void) | undefined,
+      );
+
+    return originalWrite(
+      chunk,
+      callback as ((error: Error | null | undefined) => void) | undefined,
+    );
+  };
+
+  response.end = function (...args: unknown[]) {
+    // end has these call signatures:
+    // (cb?); (chunk, cb?); (chunk, encoding, cb?)
+    // If the last parameter is a function than it's the callback
+    const argsCountBeforeCallback =
+      args.length > 0 && typeof args[args.length - 1] === "function"
+        ? args.length - 1
+        : args.length;
+    if (argsCountBeforeCallback > 0) captureChunk(args[0] as Chunk);
+
+    if (args.length == 3)
+      return originalEnd(args[0], args[1] as BufferEncoding, args[2] as () => void);
+    else if (args.length == 2) return originalEnd(args[0], args[1] as () => void);
+    else if (args.length == 1) return originalEnd(args[0]);
+    else return originalEnd();
+  };
+}
```

Comment:
This function has several problems:
- it ignores the possibility of _chunk_ being an _Uint8Array_,
- it ignores the encoding of the string when adding to the chunk array,
- the argument manipulation is opaque,
- multiple type casts make it difficult to convince the reader of correctness.

Here's my suggestion:
```typescript
function captureResponseBody(response: http.ServerResponse, chunks: Buffer[]) {
  let currentLength = 0;
  type Chunk = string | Buffer | Uint8Array;

  function captureChunk(chunk: Chunk, encoding?: BufferEncoding) {
    if (!chunk) return;
    if (config.responseBodyMaxLength <= currentLength) return;

    if (typeof chunk === "string") chunk = Buffer.from(chunk, encoding);
    chunks.push(Buffer.from(chunk));
    currentLength += chunk.length;
  }

  const originalWrite = response.write.bind(response);
  const originalEnd = response.end.bind(response);

  type WriteCB = (error: Error | null | undefined) => void;

  response.write = function (
    chunk: Chunk,
    encoding?: BufferEncoding | WriteCB,
    callback?: WriteCB,
  ) {
    if (typeof encoding === "function") {
      callback = encoding;
      encoding = undefined;
    }

    captureChunk(chunk, encoding);

    if (encoding != null) return originalWrite(chunk, encoding, callback);

    return originalWrite(chunk, callback);
  };

  type EndCB = () => void;

  response.end = function (
    chunk?: Chunk | EndCB,
    encoding?: BufferEncoding | EndCB,
    callback?: EndCB,
  ) {
    if (!chunk || typeof chunk === "function") {
      return originalEnd(chunk);
    } else if (typeof encoding === "function") {
      callback = encoding;
      encoding = undefined;
    }

    captureChunk(chunk, encoding);

    if (encoding != null) return originalEnd(chunk, encoding, callback);
    return originalEnd(chunk, callback);
  };
}
```